src/psyclone/psyir/nodes/acc_mixins.py

```diff
@@ -0,0 +1,132 @@
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors S. Valat, Inria / Lab. Jean Kuntzmann
+# Modified M. Schreiber, Univ. Grenoble Alpes / Inria / Lab. Jean Kuntzmann
+# -----------------------------------------------------------------------------
+
+''' This module contains the mixins to apply some ACC features on many
+classes.'''
+
+import abc
+
+from psyclone.psyir.nodes.reference import Reference
+
+from typing import Union
+
+
+class ACCAsyncMixin(metaclass=abc.ABCMeta):
+    '''
+    Class handling the common code to handle the async keyword on related acc
+    directives.
+
+    :param async_queue: Enable async support and attach it to the given queue.
+                        Can use False to disable, True to enable on default
+                        stream. Use int to attach to the given stream ID or
+                        use a variable Reference to say at runtime what stream
+                        to be used.
+    '''
+    def __init__(
+                self,
+                async_queue: Union[bool, int, Reference, None] = None
+            ):
+        self.async_queue = async_queue
+
+    @property
+    def async_queue(self) -> Union[bool, int, Reference, None]:
+        '''
+        :returns: whether or not async is enabled and if so, which queue this
+                  node is associated with. (True indicates the default stream.)
+                  Can use False to disable, True to enable on default stream.
+                  Int to attach to the given stream ID or use a variable
+                  Reference to say at runtime what stream to be used.
+        '''
+        return self._async_queue
+
+    @async_queue.setter
+    def async_queue(self, async_queue: Union[bool, int, Reference, None]):
+        '''
+        :param async_queue: Enable async support and attach it to the given
+                            queue. Can use False to disable, True to enable on
+                            default stream. Int to attach to the given stream
+                            ID or use a variable Reference to say at runtime
+                            what stream to be used.
+
+        :raises TypeError: if `wait_queue` is of the wrong type
+        '''
+        # check
+        if (async_queue is not None and
+           not isinstance(async_queue, (bool, Reference, int))):
+            raise TypeError(f"Invalid async_queue value, expect Reference or "
+                            f"integer or None or bool, got : {async_queue}")
+
+        # assign
+        self._async_queue = async_queue
```

Comment:
Aha, I see that we store PSyIR in a private attribute, this could be problematic (e.g. the dependency analysis will not see it, the delete symbols will not check, ...). It is also not easy to resolve (by customising: copy/signal/update_symbols/ ...) because this is a Mixin. One alternative is to make it a Clause child instead of a Mixin, but that's a big change to the current implementation (another is make it a TODO with the limitations and come back to this later), what do you think @arporter ?

Also, should we convert int to Literal?

Reply:
Yes, I think this work was begun before we had `Clause`. I agree that we should move it over. I'll take this on.